### PR TITLE
Adding cable diagnostics and PCS data-path select attribute

### PR DESCRIPTION
**Pull request by prafull-brcm**

### CODE REVIEW

**Comment by prafull-brcm** [comment]
PHY has option to select FEC mode but if NO-FEC option is selected by the user there is two data path option -
1. Re-timer pass-through mode without any processing
2. PCS IEEE data path
So we are adding attribute to select PCS IEEE data path if NO FEC option is selected by the user, if it is not selected it will follow re-timer pass-through mode.
if SAI_PORT_ATTR_USE_EXTENDED_FEC == true && SAI_PORT_ATTR_FEC_MODE_EXTENDED == SAI_PORT_FEC_MODE_EXTENDED_NONE
      Select SAI_PORT_ATTR_PCS_IEEE_DATAPATH_ENABLE True or False.

**Comment by kcudnik** [comment]
please resolve conflicts

File: inc/saiport.h [hunk 1]
```diff
@@ -106,6 +112,7 @@ typedef enum _sai_port_flow_control_mode_t
 
 /**
  * @brief Attribute data for #SAI_PORT_ATTR_INTERNAL_LOOPBACK_MODE
+ * To be deprecated, use sai_port_loopback_mode_t instead.
```
Comment (kcudnik) [inline inc/saiport.h:115]: add empty line before To line

File: inc/saiport.h [hunk 2]
```diff
@@ -210,19 +303,24 @@ typedef enum _sai_port_ptp_mode_t
...
 {
-
     /** Interface type none */
     SAI_PORT_INTERFACE_TYPE_NONE,
 
     /** Interface type CR */
     SAI_PORT_INTERFACE_TYPE_CR,
 
+    /** Interface type CR2 */
+    SAI_PORT_INTERFACE_TYPE_CR2,
```
Comment (kcudnik) [inline inc/saiport.h:313]: to be backward compatible with enum numbering you can't insert enums in between existing ones, unless you manually set their init value explicitly

File: inc/saiport.h [hunk 3]
```diff
@@ -369,6 +615,20 @@ typedef enum _sai_port_attr_t
...
+     *
+     * This attribute is applicable only for per-port, per-PG headroom model
+     * (which means SAI_BUFFER_POOL_ATTR_XOFF_SIZE is zero)
+     *
+     * For the platforms which don't have this limitation, 0 should be returned.
+     *
+     * @type sai_uint32_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_QOS_MAXIMUM_HEADROOM_SIZE,
```
Comment (kcudnik) [inline inc/saiport.h:630]: same here, you would need to number those enums explicitly or add at the end

File: inc/saiport.h [hunk 4]
```diff
@@ -698,22 +987,46 @@ typedef enum _sai_port_attr_t
 
     /**
      * @brief Internal loopback control
+     * To be deprecated, use #SAI_PORT_ATTR_LOOPBACK_MODE
```
Comment (kcudnik) [inline inc/saiport.h:990]: add empty line Before To

File: inc/saiport.h [hunk 5]
```diff
@@ -1425,57 +1778,641 @@ typedef enum _sai_port_attr_t
...
+     * @brief Attribute data for #SAI_PORT_ATTR_PRBS_RX_STATUS
+     *
+     * @type sai_port_prbs_rx_status_t
+     * @flags READ_ONLY
+     */
+    SAI_PORT_ATTR_PRBS_RX_STATUS,
+
+    /**
+     * @brief Attribute data for #SAI_PORT_ATTR_PRBS_RX_STATE
+     * Used for clear on read status/count register.
```
Comment (kcudnik) [inline inc/saiport.h:1797]: add Empty line before Used

File: inc/saiport.h (hunk 2 above)
Comment (prafull-brcm) [reply inc/saiport.h:313]: To resolve merge conflict with master I merged the change but it took it as new change. These are not my change. Please ignore this commit.

**Comment by kcudnik** [comment]
please resolve conflicts

File: inc/saiport.h (hunk 3 above)
Comment (prafull-brcm) [reply inc/saiport.h:630]: It is not my change. It is because of merge conflict it is automatically taking changes.

File: inc/saiport.h (hunk 2 above)
Comment (kcudnik) [reply inc/saiport.h:313]: then please rebase your change locally on current master, squash merge and force push to this PR

**Comment by prafull-brcm** [comment]
New Pull request created to resolve merge conflict.
https://github.com/opencomputeproject/SAI/pull/1816

**Comment by prafull-brcm** [comment]
> 

New Pull request created to resolve merge conflict.
https://github.com/opencomputeproject/SAI/pull/1816

**Pull request closed without merging**